synapse/handlers/oidc.py

```diff
@@ -1043,6 +1213,180 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
+        # This is all verified by the LogoutToken claims class, so at this
+        # point the `sid` claim exists and is a string.
+        sid: str = claims.get("sid")
+
+        # If the `sub` claim was included in the logout token, we check that it matches
+        # that it matches the right user. We can have cases where the `sub` claim is not
+        # the ID saved in database, so we let admins disable this check in config.
+        sub: Optional[str] = claims.get("sub")
+        expected_user_id: Optional[str] = None
+        if sub is not None and not self._config.backchannel_logout_ignore_sub:
+            expected_user_id = await self._store.get_user_by_external_id(
+                self.idp_id, sub
+            )
+
+        # Invalidate any running user-mapping sessions
+        self._sso_handler.revoke_mapping_sessions_for_provider_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+
+        # Invalidate any in-flight login tokens
+        await self._store.invalidate_login_tokens_by_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+
+        # Fetch any device(s) in the store associated with the session ID.
+        devices = await self._store.get_devices_by_auth_provider_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+
+        # We have no guarantee that all the devices of that session are for the same
+        # `user_id`. Hence, we have to iterate over the list of devices and log them out
+        # one by one.
+        for device in devices:
+            user_id = device["user_id"]
+            device_id = device["device_id"]
+
+            # If the user_id associated with that device/session is not the one we got
+            # out of the `sub` claim, skip that device and show log an error.
+            if expected_user_id is not None and user_id != expected_user_id:
+                logger.error(
+                    f"Received an OIDC Back-Channel Logout request "
+                    f"from issuer {self.issuer!r}, "
+                    f"for the user {expected_user_id!r} (sub: {sub!r}), "
+                    f"but with a session ID ({sid!r}) which belongs to {user_id!r}. "
+                    f"This may happen when the OIDC user mapper uses something else "
+                    f"than the `sub` claim as subject claim. "
+                    f"Set `backchannel_logout_ignore_sub` to `true` "
+                    f"in the provider config it that is the case."
+                )
+                continue
+
+            logger.info(
+                "Logging out %r (device %r) via OIDC backchannel logout (sid %r).",
+                user_id,
+                device_id,
+                sid,
+            )
+            await self._device_handler.delete_devices(user_id, [device_id])
```

Comment:
Not a huge deal, but we should probably only call this once for performance reasons.

Reply:
There technically is a code path where a single session logs out different users at the same time (like a single account on your OP branches out to two different Matrix account depending on some claim? so you can get the same session ID?), so that's why I did it this way. And since we're probably not logging out that many sessions with a single logout token, this is probably OK in terms of performance.